magento/pwa-studio
# Service Worker HTML caching bug fix

## Description

This PR addresses the issue of Venia based apps showing white screen (crashing) when a new version is released.

### Background
The Service Worker provided as part apps built using the `PWA-Studio` are built using the workbox framework. Workbox provides a bunch of utilities to ease our work of SW development/maintenance. Service worker files are built along with the client code using webpack during the build stage but are not loaded when the app loads.

Here is a timeline of what will happen when a user loads the app the first time.

```
Venia Based App Load
           |__ index.html requested
                      |__ client.************.js requested
                                 |__ Service Worker Registration
                                            |__ Service Worker Activated
                      |__ runtime.++++++++.js requested
                      |__ vendors.-----------.js requested
```

As shown above, the Service Worker is loaded and activated after the JS files are requested and loaded. Hence the first time when this happens the Service Worker was not able to handle those requests and cache their responses.

But, if the user refreshes the page or goes to a new URL in the same domain since the SW is active, it will handle those requests and if the requests have been cached in the storage, it will return them immediately instead of fetching them from the server.

The SW also caches the `index.html` file as well. `index.html` file is requested every time the user refreshes or navigates to a new URL but it is not changed often. When the file is requested by the app, the SW will return the caches HTML file from the cache and in the background request the new HTML file. After the request has been completed, the SW checks the files for changes and if it determines that the file has changed (one possibility being a new version of the app has been deployed), it sends a message to the app to show a toast asking the user to refresh.

### Problem
As mentioned above, the SW caches the HTML file for quick access. If it happens that, when the file has been returned, if it has links that are not available anymore either in cache or on the remote server, the app crashes because the network would resolve with a 404 error.

This happens to be a super edge case. This only happens when the SW has not cached all the JS files but has cached the HTML file and a new deployment is available, hence even the server does not have files anymore. This wouldn't happen if the user refreshes the page twice / gives enough time for the SW to cache the JS files.

### Steps to Reproduce
@Jordaneisenburger found the bug and has documented the steps in this [video](https://www.youtube.com/watch?v=IzNF-vv4Kow).

To test it out locally:
1. Checkout `develop` version of `pwa-studio`.
2. Build and load the app as shown in the video in different terminals. Terminal 1: `yarn build` Terminal 2: `yarn stage:venia`.
3. Load the app in a new browser instance. Make sure the cache is clean.
4. Check the cache storage section in the dev tools, you should see a cache storage instance with the word `precache` in it. It should have the HTML and JS files cached by the SW.
5. Click on the Client JS file, and delete it from the cache.
6. Now make a change in the code, as shown in the video, and perform `step 2`. This should create a new `client.[HASH].js` file.
7. Refresh the app. It should crash because the HTML file returned by the SW has reference to the old `client.[HASH].js` which is not in the cache or in the remote server.

These steps replicate a new version release because every time we release a new version, we are releasing a new version of our app's codebase which stays in `client.[HASH].js` (parts of it will be in vendors and runtime as well).

### Solution
We use the `InjectManifest` function of the `workbox-webpack-plugin@5.1.3` to inject all the assets emitted by webpack during the build process into the SW, so once the SW is activated, it will download and cache all the files mentioned in the precache manifest. 

If the build assets change, `sw.js` will change and when the active SW on the browser realizes that the precache artifacts have changed, it will gracefully delete unnecessary ones (not immediately, it is up to the SW implementation to decide when to perform this) and download the new assets.

Hence when the user goes to a new page, old assets that have been cached will be returned, and in the background, SW will update itself with all the required assets preached.

_**_Unfortunately, since this is handled by the workbox-plugin and not us, we can not let the user know if an update is available like before. I will look into way to get that working, meanwhile, this is the best solution we have._**_

To validate this, perform the same set of steps as mentioned above on this branch.

This is the 3rd version of the solution for this problem and by far the simplest solution, thanks to the [Zman](@zetlen). Here are the other 2 for reference:

https://github.com/magento/pwa-studio/tree/revanth/sw_redeploy_bug
https://github.com/magento/pwa-studio/tree/revanth/sw_redeploy_bug_v2

### Trivia
This is a major change because I had to change the signature of the [configureWebpack.js](https://github.com/magento/pwa-studio/blob/22f8c7df57f76b34a8187e75e60a85f4f0a3f76a/packages/pwa-buildpack/lib/WebpackTools/configureWebpack.js) file. This is needed because, as of the `develop` branch the Client code and the SW code are compiled using 2 different webpack compilation processes and the `InjectManifest` plugin in the SW compilation can not read the assets emitted by the client compilation. 

## Related Issue
Closes [PWA-408](https://jira.corp.magento.com/browse/PWA-408)
Closes #2351 

## Acceptance 
The app should not crash on a new version release.

### Verification Stakeholders
@zetlen 
@jimbo 
@Jordaneisenburger 
@jcalcaben 

### Verification Steps
1. Make sure the cache is cleared and load the [deployed app](https://pr-2390.pwa-venia.com/).
2. Go to the cache with the word `precache` in it. (Cache name might differ)
3. Delete the client js file from the `workbox-runtime` cache.
4. Load the app again, instead of crashing the SW should fetch the missing client js file and load the app.

You can also perform a sophisticated check like mentioned in the description, but unfortunately, it can only be done locally because it involves code changes and rebuilds.

## Checklist
* Should not change anything in `dev` mode. This is only intended for the `stage` mode because the SW is not active in the `dev/watch` mode.
* Should consult @jcalcaben to update the docs accordingly.
